### DISCUSSION TITLE
Switchable object

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#2994] [General]
Hi All,
We need a  switchable object (Say object X) in our project. 
1) Initially, the object will act as a root object, but when it is saved it will become a child object of another object (object Z).

2) Could we assign the instance created in step 1 as a root object (object X) to a child property of another object (object Z)?

3) When we call Save on the object Z, we dont object X to save again. Does the assignment operation in step 2 copy metastate properties?

Regards

**Comment by michaelcsikos** [comment]
Would it be easier to just copy the property values from the saved root object to the `Child` object and then `MarkClean()`?

**Reply by Chicagoan2016** [reply]
We would like the metastate properties copied over. Since we can't assign the child object property of a root object from UI. We are thinking the following
Inside 'Dataportal_Create' method of root object, is it possible to decide whether to call the Dataportal.Fetch<ourchildobject> or
Dataporta.Create<ourchildobject> ?

**Reply by Chicagoan2016** [reply]
It turned out we can't add a parameter to  Dataportal_Create in our root object. This is an older project targeting Csla 4.11.2.0 where we don't have access to [Create] attribute. Could we safely : ) upgrade to a version of Csla that does support [Create] attribute?

**Reply by michaelcsikos** [reply]
With 4.11.2 you can pass a `Serializable` criteria object, usually inheriting from `CriteriaBase<T>`. With your criteria class you can pass the original root object, or a DTO to copy the values.

**Reply by Chicagoan2016** [reply]
Kind of brief explanation of our use case : )
We have a root object with several 'child' objects. The object graph is working fine, there is one edge case where one of the child object's data might not be in the database and the root object needs the primary key from that data.
I created a command object to see if that particular object's data exist in the database and then pass the primary key to the root object's DataPortal_Create and based on that I call either DataPortal.CreateChild or DataPortal.Fetchchild
But for some reason, I am not a fan of this design! it assumes the UI developer will know to use the command object first.

**Reply by michaelcsikos** [reply]
In your business object you could expose a method to `GetOrCreateChildAsync(Guid? id = null)`. The business object can decide if the exists command needs to be executed first.

**Reply by Chicagoan2016** [reply]
@michaelcsikos which version of Csla supports `GetOrCreateChildASync`?

**Reply by michaelcsikos** [reply]
You can declare any factory method that you need to. In the factory method you can call the exists command and then create or fetch the object depending on the result.
